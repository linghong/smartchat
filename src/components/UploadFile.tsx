import { FC, ChangeEvent, Dispatch, SetStateAction } from 'react'

interface UploadProps {
  label: string;
  uploadError: string | null;
  setUploadError: Dispatch<SetStateAction<string | null>>;
  setSelectedFile: Dispatch<SetStateAction<File | null>>;
  fileType: string;
}

const UploadFile: FC<UploadProps> = ({ label, fileType, uploadError, setUploadError, setSelectedFile }) => {

  const handleFileChange= (e: ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files
    if(!files) {
      throw new Error("Error: Upload Files array becomes null or undefined") 
    }
    const file = files[0]
    if(file){
      const fileSize = (file.size/(1024*1024))
      if(fileSize > 1000) {
        setUploadError("Maxmium file size to upload is 1GB.")
        return
      }
      setSelectedFile(files[0])
      if(uploadError) setUploadError(null)
    }
  }
  return (
    <div className="my-2">
      <label 
        htmlFor="fileUpload"
        className="w-50 font-bold text-base mr-5">
        {label}
      </label>  
      <input
          id="fileUpload"
          type="file"
          data-testid="fileInput"
          name="uploadFile" 
          accept={fileType}
          onChange={handleFileChange}
          className="focus:border-blue-500 focus:outline-none"
      />
    </div>
  )
}

export  default UploadFile