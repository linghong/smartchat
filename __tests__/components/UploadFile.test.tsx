import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import UploadFile from '@/src/components/UploadFile'

// Mock component props
const label = "Upload Training Data: "
const fileType = ".pdf"
const uploadError = null

const setup = () => {
  const setUploadError = jest.fn()
  const setSelectedFile = jest.fn()

  render(<UploadFile label={label} fileType={fileType} uploadError={uploadError} setUploadError={setUploadError} setSelectedFile={setSelectedFile} />)
  
  const input = screen.getByTestId('fileInput') as HTMLInputElement;
  return {
    input,
    setUploadError,
    setSelectedFile,
  }
}

describe('UploadFile Component', () => {
  afterEach(() => {
    jest.clearAllMocks()
  });

  it('renders UploadFile component', () => {
    const { input } = setup()
    expect(input).toBeInTheDocument()
  })

  test('handles file change event', async() => {
    const { input, setSelectedFile, setUploadError } = setup()
    const file = new File(['dummy content'], 'example.pdf', { type: 'application/pdf' })
    await userEvent.upload(input, file)

    expect(setSelectedFile).toHaveBeenCalledWith(file)
  })

  /*test('handles file size error', () => {
    const { input, setUploadError } = setup();
    const dummyStr = 'dummy content'
    const repetitions = Math.ceil((1000 *1024*1024) / dummyStr.length)

    const largeFile = new File(['dummy content'.repeat(repetitions)], 'largeFile.pdf', { type: 'application/pdf' })
    userEvent.upload(input, largeFile)

    expect(setUploadError).toHaveBeenCalledWith("Maximum file size to upload is 200MB.")
  })*/
  
  test('UploadFile component snapshot', () => {
    const { asFragment } = render(<UploadFile label={label} fileType={fileType} uploadError={uploadError} setUploadError={jest.fn} setSelectedFile={jest.fn} />)

    expect(asFragment()).toMatchSnapshot()
  })
})