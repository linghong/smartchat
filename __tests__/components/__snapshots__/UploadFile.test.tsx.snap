// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadFile Component UploadFile component snapshot 1`] = `
<DocumentFragment>
  <div
    class="my-2"
  >
    <label
      class="w-50 font-bold text-base mr-5"
      for="fileUpload"
    >
      Upload Training Data: 
    </label>
    <input
      accept=".pdf"
      class="focus:border-blue-500 focus:outline-none"
      data-testid="fileInput"
      id="fileUpload"
      name="uploadFile"
      type="file"
    />
  </div>
</DocumentFragment>
`;
